sympy/core/power.py

```diff
@@ -181,6 +181,35 @@ def integer_log(y, x):
     return e, r == 0 and y == 1
 
 
+def binpow(b, e):
```

Comment:
It seems that Python implements this algorithm [in C](https://github.com/python/cpython/blob/953e733e6d0bc02fde0e4f808259218a95a3b902/Objects/longobject.c#L4328-L4360). That would be more efficient.

Reply:
Aha! Thank you for the link. You know, I was wondering if this might have been the case, but didn't know how to investigate it.

Well, my question now is whether there's a good way to bring Python's internal powering algorithm to apply to powers of user-defined classes. In other words, suppose I'm defining `class Foo` and I've defined `Foo.__mul__`. Is there a way to then tell Python that, when it sees

```python
f = Foo()
g = f ** 3
```

it is to use its internal binary powering algorithm, while calling on `Foo.__mul__` each time it needs to actually multiply two `Foo`s together in the process? If so, that would be great!

Reply:
I think that the C version is only for `int`. This probably does not belong in core though. I don't think that core is a good place for this (especially if it isn't used there) but maybe `sympy.external` can do for now.